WordPress/wordpress-develop
# Add directory support to `WP_Filesystem_Direct::move()`.

This PR adds directory support to `WP_Filesystem_Direct::move()` for consistency with the other `WP_Filesystem_*` classes.

This includes:
- Updates to `WP_Filesystem_Direct::move()`.
- A new global function: `wp_opcache_invalidate_directory()`, to recursively call `wp_opcache_invalidate()`.
- A new wrapper function: `move_dir()`, for extender convenience and more informative failure reasons.
- Tests for `move_dir()`.
- Tests for `wp_opcache_invalidate_directory()`.

### Testing instructions
#### Steps to test
1. Ensure that `WP_DEBUG` and `WP_DEBUG_LOG` are set to `true`.
2. Change [this line](https://github.com/costdev/wordpress-develop/blob/trunk/src/wp-admin/includes/class-wp-upgrader.php#L597) to:

```php
$result = move_dir( $source, $remote_destination, true );
```
3. Navigate to `Dashboard > Updates` and update one or more plugins/themes.
4. Navigate to `Plugins > Installed plugins` and update one or more plugins.
5. Navigate to `Appearance > Themes ` and update some themes.

#### Expected results

All updates should complete as normal, with no errors/warnings/notices in `wp-content/debug.log`.

Trac ticket: https://core.trac.wordpress.org/ticket/57375
